docs/ffi/ffi.md

```diff
@@ -0,0 +1,467 @@
+
+JAX's FFI support is provided in two parts:
+
+1. A header-only C++ library from XLA which is packaged as part of JAX as of v0.4.29 or available from the [openxla/xla](https://github.com/openxla/xla) project, and
+2. A Python front end, available in the `jax.extend.ffi` submodule.
+
+In this tutorial we demonstrate the use of both of these components using a simple example, and then go on to discuss some lower-level extensions for more complicated use cases.
+We start by presenting the FFI on CPU, and discuss generalizations to GPU or multi-device environments below.
+
+This tutorial comes with two supplementary files:
+
+* [`rms_norm.cc`](rms_norm.cc), which includes all the backend code, and
+* [`CMakeLists.txt`](CMakeLists.txt), which tells [CMake](https://cmake.org) how to build the code.
+
+## A simple example
+
+To demonstrate the use of the FFI interface, we will implement a simple "root-mean-square (RMS)" normalization function.
+RMS normalization takes an array $x$ with shape $(N,)$ and returns
+
+$$
+y_n = \frac{x_n}{\sqrt{\frac{1}{N}\sum_{n=1}^N {x_n}^2 + \epsilon}}
+$$
+
+where $\epsilon$ is a tuning parameter used for numerical stability.
+
+This is a somewhat silly example, because it can be easily implemented using JAX as follows:
+
+```{code-cell} ipython3
+import jax
+import jax.numpy as jnp
+
+
+def rms_norm_ref(x, eps=1e-5):
+  scale = jnp.sqrt(jnp.mean(jnp.square(x), axis=-1, keepdims=True) + eps)
+  return x / scale
+```
+
+But, it's just non-trivial enough to be useful for demonstrating some key details of the FFI, while still being straightforward to understand.
+We will use this reference implementation to test our FFI version below.
+
+## Backend code
+
+To begin with, we need an implementation of RMS normalization in C++ that we will expose using the FFI.
+This isn't meant to be particularly performant, but you could imagine that if you had some new better implementation of RMS normalization in a C++ library, it might have an interface like the following.
+So, here's a simple implementation of RMS normalization in C++:
+
+```c++
+#include <cmath>
+#include <cstdint>
+
+float ComputeRmsNorm(float eps, int64_t size, const float *x, float *y) {
+  float sm = 0.0f;
+  for (int64_t n = 0; n < size; ++n) {
+    sm += x[n] * x[n];
+  }
+  float scale = 1.0f / std::sqrt(sm / float(size) + eps);
+  for (int64_t n = 0; n < size; ++n) {
+    y[n] = x[n] * scale;
+  }
+  return scale;
+}
+```
+
+and, for our example, this is the function that we want to expose to JAX via the FFI.
+
+### C++ interface
+
+To expose our library function to JAX and XLA, we need to write a thin wrapper using the APIs provided by the header-only library in the [`xla/ffi/api`](https://github.com/openxla/xla/tree/main/xla/ffi/api) directory of the [XLA project](https://github.com/openxla/xla).
+For more information about this interface, take a look at [the XLA custom call documentation](https://openxla.org/xla/custom_call).
```

Comment:
Is there documentation for the `Ffi::Ffi::Bind`? Last time I looked a few months ago, I could not find any.

Reply:
I'll add more details to https://openxla.org/xla/custom_call